File: battle/data/MaskEntity.java
package common.battle.data;

import common.pack.Identifier;
import common.util.Animable;
import common.util.BattleStatic;
import common.util.Data;
import common.util.Data.Proc;
import common.util.anim.AnimU;
import common.util.anim.AnimU.UType;
import common.util.pack.Soul;
import common.util.unit.Trait;

import java.util.ArrayList;

public interface MaskEntity extends BattleStatic {

	int allAtk();

	int getAbi();

	Proc getAllProc();

	/**
	 * get the attack animation length
	 */
	default int getAnimLen() {
		return getPack().anim.getAtkLen();
	}

	int getAtkCount();

	int getAtkLoop();

	MaskAtk getAtkModel(int ind);

	MaskAtk[] getAtks();

	default AtkDataModel[] getSpAtks() {
		return null;
	}

	Identifier<Soul> getDeathAnim();

	ArrayList<Trait> getTraits();

	int getHb();

	int getHp();

	/**
	 * get the attack period
	 */
	int getItv();

	/**
	 * get the Enemy/Form this data represents
	 */
	Animable<AnimU<?>, UType> getPack();

	int getPost();

	Proc getProc();

	int getRange();

	MaskAtk getRepAtk();

	default AtkDataModel getResurrection() {
		return null;
	}

	default AtkDataModel getRevenge() {
		return null;
	}

	default AtkDataModel getCounter() {
		return null;
	}

	default AtkDataModel getGouge() {
		return null;
	}

	default AtkDataModel getResurface() {
		return null;
	}

	default AtkDataModel getRevive() {
		return null;
	}

	int getSpeed();

	int getWill();

	/**
	 * get waiting time
	 */
	int getTBA();

	default int getTouch() {
		return Data.TCH_N;
	}

	int getWidth();

	boolean isLD();

	default boolean isLD(int ind) {
		return isLD();
	}

	boolean isOmni();

	default boolean isOmni(int ind) {
		return isOmni();
	}

	boolean isRange();

	int[][] rawAtkData();

	int touchBase();

}

File: battle/data/MaskEntity.kt
package common.battle.data

import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.PackData
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.Animable
import common.util.BattleStatic
import common.util.Data
import common.util.Data.Proc
import common.util.anim.AnimU
import common.util.anim.AnimU.UType
import common.util.pack.Soul
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter

interface MaskEntity : BattleStatic {
    fun allAtk(): Int
    fun getAbi(): Int
    fun getAllProc(): Proc?

    /** get the attack animation length  */
    fun getAnimLen(): Int {
        return getPack().anim.getAtkLen()
    }

    fun getAtkCount(): Int
    fun getAtkLoop(): Int
    fun getAtkModel(ind: Int): MaskAtk?
    fun getDeathAnim(): PackData.Identifier<Soul>?
    fun getHb(): Int
    fun getHp(): Int

    /** get the attack period  */
    fun getItv(): Int

    /** get the Enemy/Form this data represents  */
    fun getPack(): Animable<AnimU<*>?, UType?>?
    fun getPost(): Int
    fun getProc(): Proc?
    fun getRange(): Int
    fun getRepAtk(): MaskAtk?
    fun getResurrection(): AtkDataModel? {
        return null
    }

    fun getRevenge(): AtkDataModel? {
        return null
    }

    fun getShield(): Int
    fun getSpeed(): Int

    /** get waiting time  */
    fun getTBA(): Int
    fun getTouch(): Int {
        return Data.Companion.TCH_N
    }

    fun getType(): Int
    fun getWidth(): Int
    fun isLD(): Boolean
    fun isOmni(): Boolean
    fun isRange(): Boolean
    fun rawAtkData(): Array<IntArray?>
    fun touchBase(): Int
}
